src/Passes/Source/Validator/Validator.cpp

```diff
@@ -48,37 +71,94 @@ namespace quantum
         }
         catch (std::exception const& e)
         {
-            llvm::errs() << "; Fatal error: " << e.what() << "\n";
+            if (logger_)
+            {
+                logger_->error("Fatal error: " + static_cast<std::string>(e.what()));
+            }
+
+            saveLogsToFileIfNeeded();
             return false;
         }
-        return true;
-    }
 
-    void Validator::setModulePassManager(llvm::ModulePassManager&& manager)
-    {
-        module_pass_manager_ = std::move(manager);
+        saveLogsToFileIfNeeded();
+        return true;
     }
 
     llvm::PassBuilder& Validator::passBuilder()
     {
         return *pass_builder_;
     }
+
     llvm::LoopAnalysisManager& Validator::loopAnalysisManager()
     {
         return loop_analysis_manager_;
     }
+
     llvm::FunctionAnalysisManager& Validator::functionAnalysisManager()
     {
         return function_analysis_manager_;
     }
+
     llvm::CGSCCAnalysisManager& Validator::gsccAnalysisManager()
     {
         return gscc_analysis_manager_;
     }
+
     llvm::ModuleAnalysisManager& Validator::moduleAnalysisManager()
     {
         return module_analysis_manager_;
     }
 
+    void Validator::saveLogsToFileIfNeeded()
```

Comment:
I would expect such a helper to be in the logger rather than part of the validator.

Reply:
The idea is that this will write a validation report even if it just writing a log to JSON format atm. Hence it is meant to be beyond just writing a log. I've updated the function name to reflect this.